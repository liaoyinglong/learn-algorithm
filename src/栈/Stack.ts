export class Stack<T> {
  private items = [] as T[];
  // 入栈
  push(item: T) {
    this.items.push(item);
  }
  // 出栈
  pop(): T | undefined {
    return this.items.pop();
  }
  // 读栈顶 即最后一个
  peek(): T | undefined {
    return this.items[this.items.length - 1];
  }

  // 判空
  isEmpty(): boolean {
    return !this.items.length;
  }
  // 读栈长度
  size(): number {
    return this.items.length;
  }

  // 清空栈
  clear(): void {
    this.items = [];
  }

  static from<T>(arr: T[]) {
    const t = new Stack<T>();
    arr.forEach((item) => t.push(item));
    return t;
  }
}
