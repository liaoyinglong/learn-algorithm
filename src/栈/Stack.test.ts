import { Stack } from "src/æ ˆ/Stack";

describe("Stack", () => {
  let t: Stack<string>;

  beforeEach(() => {
    t = Stack.from(["a", "b", "c"]);
  });

  it("pop method", () => {
    t.pop();
    expect(t).toMatchSnapshot();
    expect(t.size()).toBe(2);
    expect(t.isEmpty()).toBe(false);
  });

  it("peek method", () => {
    expect(t.peek()).toBe("c");
  });

  it("clear method", () => {
    t.clear();
    expect(t).toMatchSnapshot();
    expect(t.size()).toBe(0);
    expect(t.isEmpty()).toBe(true);
  });
});
