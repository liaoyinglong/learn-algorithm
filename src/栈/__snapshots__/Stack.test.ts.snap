// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack clear method 1`] = `
Stack {
  "items": Array [],
}
`;

exports[`Stack pop method 1`] = `
Stack {
  "items": Array [
    "a",
    "b",
  ],
}
`;
